### ISSUE TITLE
Type match in the "transparent scope" of opaque types

### ISSUE THREAD

**Issue by ftucky** [lampepfl/dotty#17580]
## Compiler version

3.3.0-RC3

## Minimized code

```Scala
object Foo:
	opaque type T = Double

	type Match[A] = A match
		case Int   => String
		case T     => String

val a: Foo.Match[String] = 0
```

## Output

```
Found:    (0 : Int)
Required: Playground.Foo.Match[String]

Note: a match type could not be fully reduced:

  trying to reduce  Playground.Foo.Match[String]
  failed since selector  String
  does not match  case Playground.Foo.T => String
  and cannot be shown to be disjoint from it either.
```

`Foo.Match[String]` is considered a valid type, which it is not ( `String` is neither a `Int` nor a `T` )

## Expectation

The compiler should reject `Match[String]` as invalid like this :

```
Match type reduction failed since selector  String
matches none of the cases

    case Int => String
    case Playground.Foo.T => String
```

## Observations

1/ When `T` is *not* opaque, `Match[String]` is correctly rejected.

```
object Foo:
	type T = Double

	type Match[A] = A match
		case Int   => String
		case T     => String

val a: Foo.Match[String] =  ??? // <-- Correctly rejected
```

2/ When `T` is not defined in an object, `Match[String]` is correctly rejected. 
```
opaque type T = Double

type Match[A] = A match
	case Int   => String
	case T     => String

val a: Match[String] =  ???  // <- Correctly rejected.
```

3/ When Match is used in the `object Foo`, it is also correctly rejected
```
object Foo:
	type T = Double

	type Match[A] = A match
		case Int   => String
		case T     => String

        val a: Match[String] =  ??? // <-- Correctly rejected
```

## Analysis

The issue seems to be related the scope where `opaque type T` is *not* opaque.
The match type is defined inside this scope, but used outside of it.

As `T` appears in the match type definition, one would expect `T` *not to be opaque at that point.
However compiler seems to take the opacity at *usage place* into account.
This is not consistent as the programmer using the match type is probably not even aware of the opaque type.

**Comment by odersky** [comment]
I think this is by design. It's technically next to impossible to resolve match types in the scope of their definition. Even if we could that would mean that the types of opaques are visible in some way at the use site, which is not what we want.

**Comment by ftucky** [comment]
Yes, you are right. Point taken.


Consider the following code:
```
object Foo:
	opaque type T = Double
	type Match[A] = A match { case T => String }

val a : Foo.Match[Double] =  ""
```

Reports : 
>   trying to reduce  Foo.Match[Double]
>   failed since selector  Double
>   does not match  case Foo.T => String
>   and **cannot be shown to be disjoint from it either.**

as if `Foo.T` was declared but undefined.
Here `Foo.T` *is* defined as being **opaque**.
Don't we expect to see it not to match anything (but itself), not even a `Double` ?

Otherwise, how may *opaque types* be used in *type matches* at all ?

**Comment by odersky** [comment]
> as if Foo.T was declared but undefined.

Well, it is defined but unknown.

> Otherwise, how may opaque types be used in type matches at all ?

I guess only by giving literally the opaque type:

```scala
Foo.Match[Foo.T]
```

**Comment by sjrd** [comment]
As a reminder,
```scala
opaque type T = Double
```
really means
```scala
opaque type T >: Nothing <: Any = Double
```
and as soon as you "leave" its defining scope, the `= Double` gets lost (on purpose), so all you have left is
```scala
type T >: Nothing <: Any
```
and that is clearly not disjoint from `String`.

**Comment by t9dupuy** [comment]
> It works if you invert the cases, also it works for subtypes:

Not really: if you invert the cases you invert the problem as type `Int` is not matched anymore

```scala
object Foo:
  opaque type T = Double

  type Match[A] = A match
    case T   => String
    case Int => String

val a: Foo.Match[Int] = 0
```

will result in 

```
Required: Foo.Match[Int]

Note: a match type could not be fully reduced:

    trying to reduce  Foo.Match[Int]
    failed since selector  Int
    does not match  case Foo.T => String
    and cannot be shown to be disjoint from it either.
```
